src/main.clj

```diff
@@ -1,2 +1,314 @@
+((defn aoc1-part2
+   [file]
+   (with-open [rdr (clojure.java.io/reader file)]
+     (:count
+       (reduce
+         (fn [report depth]
+           (let [aggregated-depth (reduce (fn [x y] (+ x (Integer/parseInt y))) 0 (vec depth))]
+             (if (< (:prev-depth report) aggregated-depth)
+               ;  If depth has increased, increase the count.
+               (update (assoc report :prev-depth aggregated-depth) :count inc)
+               ;  Just update previous depth in other case.
+               (assoc report :prev-depth aggregated-depth))))
+         ; Set previous depth as maximum for initial value.
+         {:prev-depth Integer/MAX_VALUE :count 0}
+         ; Feed sequence of readings to reduce.
+         (partition 3 1 (line-seq rdr))))))
+ "/Users/a/w/4clojure/aoc1-input.txt")
+
+; https://adventofcode.com/2021/day/2#part1
+((defn aoc2-part1
+   [file]
+   (with-open [rdr (clojure.java.io/reader file)]
+     (let
+       [position
+        (reduce
+          (fn [position cmd]
+            (let [split-cmd (clojure.string/split cmd #" ")
+                  units (Integer/parseInt (nth split-cmd 1))
+                  command (nth split-cmd 0)]
+              (case command
+                "forward" (update position :horizontal + units)
+                "down" (update position :depth + units)
+                "up" (update position :depth - units))))
+          ; Starting position is 0
+          {:horizontal 0 :depth 0}
+          ; Feed sequence of readings to reduce.
+          (line-seq rdr))]
+       (* (:horizontal position) (:depth position)))))
+ "/Users/a/w/4clojure/aoc2-input.txt")
+
+
+; https://adventofcode.com/2021/day/2#part2
+((defn aoc2-part2
+   [file]
+   (with-open [rdr (clojure.java.io/reader file)]
+     (let
+       [position
+        (reduce
+          (fn [position cmd]
+            (let [split-cmd (clojure.string/split cmd #" ")
+                  units (Integer/parseInt (nth split-cmd 1))
+                  command (nth split-cmd 0)]
+              (case command
+                "forward" (as-> position p
+                                (update p :horizontal + units)
+                                (update p :depth + (* units (:aim p))))
+                "down" (update position :aim + units)
+                "up" (update position :aim - units))))
+          ; Starting position is 0
+          {:horizontal 0 :depth 0 :aim 0}
+          ; Feed sequence of readings to reduce.
+          (line-seq rdr))]
+       (* (:horizontal position) (:depth position)))))
+ "/Users/a/w/4clojure/aoc2-input.txt")
+
+; Mars Rover - https://github.com/priyaaank/MarsRover
+; QQQ: How to achieve extensibility in Clojure?
+; The way I see it, all inter-method calls/reduce are deeply coupled.
+; And if I were to look at this after a few months, I won't be able to make head/tail out of this :(
+(defn parse-input
+  [[x y d]]
+  [(Integer/parseInt x) (Integer/parseInt y) (keyword d)])
```

Comment:
Prefer maps over vectors for modelling associative data. A rover has an x-coordinate, a y-coordinate and a direction. It might look like this:
```
{:x 5
 :y 6
 :direction :N}
```
With the vector modelling, one has to carry around the tacit knowledge that the first item is the x-coordinate, the second item is the y-coordinate and the third item is the direction. It's easier to use a map and give them names directly.
When using a vector or a sequence, ask yourself if the order of elements should matter. If not, you may be better off using a map or a set.